airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/dest_state_lifecycle_manager/DestStreamStateLifecycleManager.java

```diff
@@ -0,0 +1,111 @@
+ * NOT guarantee that state messages of different streams will be output in the order in which they
+ * were received. State messages across streams will be emitted in alphabetical order (primary sort
+ * on namespace, secondary on name).
+ */
+public class DestStreamStateLifecycleManager implements DestStateLifecycleManager {
+
+  private final Map<StreamDescriptor, AirbyteMessage> streamToLastPendingState;
+  private final Map<StreamDescriptor, AirbyteMessage> streamToLastFlushedState;
+  private final Map<StreamDescriptor, AirbyteMessage> streamToLastCommittedState;
+
+  public DestStreamStateLifecycleManager() {
+    streamToLastPendingState = new HashMap<>();
+    streamToLastFlushedState = new HashMap<>();
+    streamToLastCommittedState = new HashMap<>();
+  }
+
+  @Override
+  public void addState(final AirbyteMessage message) {
+    Preconditions.checkArgument(message.getState().getStateType() == AirbyteStateType.STREAM);
+    streamToLastPendingState.put(message.getState().getStream().getStreamDescriptor(), message);
+  }
+
+  @VisibleForTesting
+  Queue<AirbyteMessage> listPending() {
+    return listStatesInOrder(streamToLastPendingState);
+  }
+
+  @Override
+  public void markPendingAsFlushed() {
+    moveToNextPhase(streamToLastPendingState, streamToLastFlushedState);
+  }
+
+  @Override
+  public Queue<AirbyteMessage> listFlushed() {
+    return listStatesInOrder(streamToLastFlushedState);
+  }
+
+  @Override
+  public void markFlushedAsCommitted() {
+    moveToNextPhase(streamToLastFlushedState, streamToLastCommittedState);
+  }
+
+  @Override
+  public Queue<AirbyteMessage> listCommitted() {
+    return listStatesInOrder(streamToLastCommittedState);
+  }
+
+  /**
+   * Lists out the states in the stream to state maps. Guarantees a deterministic sort order, which is
+   * handy because we are going from a map (unsorted) to a queue. The sort order primary sort on
+   * namespace (with null at the top) followed by secondary sort on name. This maps onto the pretty
+   * common order that we list streams elsewhere.
+   *
+   * @param streamToState - map of stream descriptor to its last state
+   * @return queue with the states ordered per the sort mentioned above
+   */
+  private static Queue<AirbyteMessage> listStatesInOrder(final Map<StreamDescriptor, AirbyteMessage> streamToState) {
+    return streamToState
+        .entrySet()
+        .stream()
+        // typically, we support by namespace and then stream name, so we retain that pattern here.
+        .sorted(Comparator
+            .comparing(
+                (Function<Entry<StreamDescriptor, AirbyteMessage>, String>) entry -> entry.getKey().getNamespace(),
+                Comparator.nullsFirst(Comparator.naturalOrder())) // namespace is allowed to be null
+            .thenComparing(entry -> entry.getKey().getName()))
+        .map(Entry::getValue)
+        .collect(Collectors.toCollection(LinkedList::new));
```

Comment:
Nitpick. When the size of the list is known and the list is immutable, `ArrayList` has a lower memory overhead than `LinkedList`.

Reply:
`ArrayList` doesn't implement `Queue` though.